src/v/utils/available_promise.h

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+#include "seastarx.h"
+
+#include <seastar/core/future.hh>
+
+template<typename T>
+class available_promise {
+public:
+    ss::future<T> get_future() { return _promise.get_future(); }
+
+    void set_value(T&& value) {
+        _available = true;
+        _promise.set_value(std::move(value));
```

Comment:
> you can move the promise value itself after you create the future. in that case, avialable() wouldn't be necessary when do_end_txn returns if the return value were richer. you'd set the outcome unconditionally.

as I understand if I move a promise into `do_end_txn` I don't be able to set it after `do_end_txn` returns control flow because the promise was moved

> you'd set the outcome unconditionally

I don't understand this, I need to set the promise after the `do_end_txn` as a fail safe mechanism, in case the method itself hasn't done it so I have to check if it's not available before doing so

Reply:
The idea was to not pass a promise into `do_end_txn` at all. Instead, return an equivalent to: `std::pair<ss::future<checked<cluster::tm_transaction, tx_errc>>, tx_errc>`

and the caller would set the promise based on ret.second (because it has the promise it would have originally been passing in as a parameter), and it can wait on ret.first instead of waiting on `do_end_txn`.

cleaning this up isn't necessarily for the PR but it is some real complexity that looks like it could be eliminated.

Reply:
created a follow up task - https://github.com/vectorizedio/redpanda/issues/1566